File: app/src/main/java/com/example/lenovo/englishstudy/fragment/HomeFragment.java
package com.example.lenovo.englishstudy.fragment;

import android.app.Activity;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.view.ViewPager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ScrollView;
import android.widget.Toast;

import com.example.lenovo.englishstudy.R;
import com.example.lenovo.englishstudy.Util.GetRequest_Interface;
import com.example.lenovo.englishstudy.bean.ArticleList;
import com.example.lenovo.englishstudy.bean.SexagenaryCycle;
import com.example.lenovo.englishstudy.pullextend.PullToRefreshLinearLayout;
import com.example.lenovo.englishstudy.viewPageCard.CardItem;
import com.example.lenovo.englishstudy.viewPageCard.CardPagerAdapter;
import com.example.lenovo.englishstudy.viewPageCard.ShadowTransformer;
import com.jakewharton.retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class HomeFragment extends Fragment {

    private ViewPager mViewPager;
    private Activity activity;
    private CardPagerAdapter mCardAdapter;
    private ShadowTransformer mCardShadowTransformer;
    @BindView(R.id.lin_refresh)
    PullToRefreshLinearLayout linRefresh;
    Handler handler = new Handler();
    @BindView(R.id.sv_scrollView)
    ScrollView scrollView;
    private String sexagenaryCycleYear;
    private String lunarCalendar;
    SexagenaryCycle mSexagenaryCycle;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.homefragment, container, false);
        ButterKnife.bind(view);
        scrollView = view.findViewById(R.id.sv_scrollView);
        mViewPager = view.findViewById(R.id.vp_WordCard);
        linRefresh = view.findViewById(R.id.lin_refresh);
        requestArticleList();
        initEvent();
        return view;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
    }

    private void initEvent() {
        // linRefresh.hideFooterView();
        linRefresh.setOnHeaderRefreshListener(new PullToRefreshLinearLayout.OnHeaderRefreshListener() {
            @Override
            public void onHeaderRefresh(PullToRefreshLinearLayout view) {
                handler.postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        linRefresh.onHeaderRefreshComplete();
                        // linRefresh.showFooterView();
                        requestArticleList();
                    }
                }, 2000);
            }
        });

        linRefresh.setOnFooterRefreshListener(new PullToRefreshLinearLayout.OnFooterRefreshListener() {
            @Override
            public void onFooterRefresh(PullToRefreshLinearLayout view) {
                handler.postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        linRefresh.onFooterRefreshComplete();
                        requestArticleList();
                    }
                }, 2000);
            }
        });
    }

    public void requestArticleList() {
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl("http://47.102.206.19:8080/")
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build();

        GetRequest_Interface request = retrofit.create(GetRequest_Interface.class);

        Observable<ArticleList> observable = request.getArticleList();

        observable.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<ArticleList>() {
                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onNext(ArticleList articleList) {
                        showArticleList(articleList);
                    }

                    @Override
                    public void onError(Throwable e) {
                        Toast.makeText(getContext(), "获取单词联想失败1", Toast.LENGTH_SHORT).show();
                    }

                    @Override
                    public void onComplete() {
                    }
                });
    }

    public void showArticleList(ArticleList articleList) {
        if (articleList != null) {
            activity = getActivity();
            mCardAdapter = new CardPagerAdapter(activity);
            for (int i = 0; i < articleList.getData().getTotal(); i++) {
                mCardAdapter.addCardItem(new CardItem(articleList.getData().getList().get(i)));
            }
            mCardShadowTransformer = new ShadowTransformer(mViewPager, mCardAdapter);
            mCardShadowTransformer.enableScaling(true);
            mViewPager.setAdapter(mCardAdapter);
            mViewPager.setPageTransformer(false, mCardShadowTransformer);
            mViewPager.setOffscreenPageLimit(3);
        }
    }
}

File: app/src/main/java/com/example/lenovo/englishstudy/fragment/HomeFragment.kt
package com.example.lenovo.englishstudy.fragment

import android.app.Activity
import android.os.Bundle
import android.os.Handler
import android.support.v4.app.Fragment
import android.support.v4.view.MenuItemCompat
import android.support.v4.view.ViewPager
import android.support.v7.widget.Toolbar
import android.view.*
import android.widget.ScrollView
import android.widget.SearchView
import android.widget.Toast
import com.example.lenovo.englishstudy.R
import com.example.lenovo.englishstudy.Util.GetRequest_Interface
import com.example.lenovo.englishstudy.bean.ArticleList
import com.example.lenovo.englishstudy.pullextend.PullToRefreshLinearLayout
import com.example.lenovo.englishstudy.viewPageCard.CardItem
import com.example.lenovo.englishstudy.viewPageCard.CardPagerAdapter
import com.example.lenovo.englishstudy.viewPageCard.ShadowTransformer
import com.jakewharton.retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class HomeFragment : Fragment() {

    private var activity: Activity? = null
    private var mCardAdapter: CardPagerAdapter? = null
    private var mCardShadowTransformer: ShadowTransformer? = null
    private var handler = Handler()
    private lateinit var linRefresh: PullToRefreshLinearLayout
    private lateinit var mViewPager: ViewPager
    private lateinit var scrollView: ScrollView
    private lateinit var toolbar: Toolbar
    private lateinit var mSearchView: SearchView

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.homefragment, container, false)
        scrollView = view.findViewById(R.id.sv_scrollView)
        mViewPager = view.findViewById(R.id.vp_WordCard)
        linRefresh = view.findViewById(R.id.lin_refresh)
        toolbar = view.findViewById(R.id.tb_home_fragment)

        toolbar.inflateMenu(R.menu.toolbar_homefragment)
        toolbar.setOnMenuItemClickListener {
            when (it.itemId) {
                R.id.homefragment_message ->
                    Toast.makeText(context, "show message", Toast.LENGTH_SHORT).show()
                R.id.homefragment_history ->
                    Toast.makeText(context, "show history", Toast.LENGTH_SHORT).show()
            }
            true
        }

        requestArticleList()
        initEvent()
        return view
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        inflater?.inflate(R.menu.toolbar_homefragment, menu)
        val searchItem: MenuItem = menu!!.findItem(R.id.homefragment_search)
        mSearchView = MenuItemCompat.getActionView(searchItem) as SearchView
        mSearchView.setIconifiedByDefault(false)
        mSearchView.isSubmitButtonEnabled = true
        mSearchView.setOnQueryTextListener(object : android.support.v7.widget.SearchView.OnQueryTextListener, SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(p0: String?): Boolean {
                return false
            }

            override fun onQueryTextChange(p0: String?): Boolean {
                Toast.makeText(context, "搜索: $p0", Toast.LENGTH_LONG).show()
                return false
            }
        })
    }

    private fun initEvent() {
        // linRefresh.hideFooterView();
        linRefresh.setOnHeaderRefreshListener {
            handler.postDelayed({
                linRefresh.onHeaderRefreshComplete()
                // linRefresh.showFooterView();
                requestArticleList()
            }, 2000)
        }

        linRefresh.setOnFooterRefreshListener {
            handler.postDelayed({
                linRefresh.onFooterRefreshComplete()
                requestArticleList()
            }, 2000)
        }
    }

    fun requestArticleList() {
        val retrofit = Retrofit.Builder()
                .baseUrl("http://47.102.206.19:8080/")
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build()

        val request = retrofit.create(GetRequest_Interface::class.java)

        val observable = request.articleList

        observable.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(object : Observer<ArticleList> {
                    override fun onSubscribe(d: Disposable) {

                    }

                    override fun onNext(articleList: ArticleList) {
                        showArticleList(articleList)
                    }

                    override fun onError(e: Throwable) {
                        Toast.makeText(context, "获取单词联想失败1", Toast.LENGTH_SHORT).show()
                    }

                    override fun onComplete() {}
                })
    }

    fun showArticleList(articleList: ArticleList?) {
        if (articleList != null) {
            activity = getActivity()
            mCardAdapter = CardPagerAdapter(activity)
            for (i in 0 until articleList.data.total) {
                mCardAdapter!!.addCardItem(CardItem(articleList.data.list[i]))
            }
            mCardShadowTransformer = ShadowTransformer(mViewPager, mCardAdapter)
            mCardShadowTransformer!!.enableScaling(true)
            mViewPager.adapter = mCardAdapter
            mViewPager.setPageTransformer(false, mCardShadowTransformer)
            mViewPager.offscreenPageLimit = 3
        }
    }
}
